solidity/contracts/middleware/libs/InterchainAccountMessage.sol

```diff
@@ -107,27 +117,84 @@ library InterchainAccountMessage {
         CallLib.Call[] calldata _calls,
         bytes32 _userSalt
     ) internal pure returns (bytes memory) {
-        return abi.encode(_owner, _ism, _calls, _userSalt);
+        bytes memory prefix = abi.encodePacked(
+            MessageType.CALLS,
+            _owner,
+            _ism,
+            _userSalt
+        );
+        bytes memory suffix = abi.encode(_calls);
+        return bytes.concat(prefix, suffix);
     }
 
     /**
-     * @notice Parses and returns the calls from the provided message
-     * @param _message The interchain account message
-     * @return The array of calls
+     * @notice Returns formatted (packed) InterchainAccountMessage
+     * @dev This function should only be used in memory message construction.
+     * @param _owner The owner of the interchain account
+     * @param _ism The address of the remote ISM
+     * @return Formatted message body
      */
-    function decode(
-        bytes calldata _message
-    ) internal pure returns (bytes32, bytes32, CallLib.Call[] memory, bytes32) {
+    function encodeCommitment(
+        bytes32 _owner,
+        bytes32 _ism,
+        bytes32 _commitment,
+        bytes32 _userSalt
+    ) internal pure returns (bytes memory) {
         return
-            abi.decode(_message, (bytes32, bytes32, CallLib.Call[], bytes32));
+            abi.encodePacked(
+                MessageType.COMMITMENT,
+                _owner,
+                _ism,
+                _userSalt,
+                _commitment
+            );
+    }
+
+    function encodeReveal(
+        bytes32 _ism,
+        bytes32 _commitment
+    ) internal pure returns (bytes memory) {
+        return abi.encodePacked(MessageType.REVEAL, _ism, _commitment);
+    }
+
+    function messageType(
+        bytes calldata _message
+    ) internal pure returns (MessageType) {
+        return MessageType(uint8(bytes1(_message[0:1])));
+    }
+
+    function owner(bytes calldata _message) internal pure returns (bytes32) {
+        return bytes32(_message[1:33]);
     }
 
     /**
      * @notice Parses and returns the ISM address from the provided message
      * @param _message The interchain account message
      * @return The ISM encoded in the message
      */
-    function ism(bytes calldata _message) internal pure returns (address) {
-        return address(bytes20(_message[44:64]));
+    function ism(bytes calldata _message) internal pure returns (bytes32) {
+        if (messageType(_message) == MessageType.REVEAL) {
+            return bytes32(_message[1:33]);
+        }
+        return bytes32(_message[33:65]);
```

Comment:
I dont love branching in here
could we have separate codecs for each message type?

Reply:
Updated a couple functions to take message type as a param. This doesn't fix branching, but I think using a parameter is cleaner than using a different function signature for every different kind of message. If we have more message types in the future, using different functions could become unwieldly.